src/Processors/QueryPlan/Optimizations/tryUpdateQueryConditionCache.cpp

```diff
@@ -0,0 +1,40 @@
+#include <Processors/QueryPlan/Optimizations/Optimizations.h>
+#include <Processors/QueryPlan/FilterStep.h>
+#include <Processors/QueryPlan/ReadFromMergeTree.h>
+#include <Storages/VirtualColumnUtils.h>
+
+namespace DB::QueryPlanOptimizations
+{
+
+void tryUpdateQueryConditionCache(const QueryPlanOptimizationSettings & optimization_settings, const Stack & stack)
+{
+    if (!optimization_settings.enable_writes_to_query_condition_cache)
+        return;
+
+    const auto & frame = stack.back();
+
+    auto * read_from_merge_tree = dynamic_cast<ReadFromMergeTree *>(frame.node->step.get());
+    if (!read_from_merge_tree)
+        return;
+
+    const auto & query_info = read_from_merge_tree->getQueryInfo();
+    auto filter_dag = query_info.filter_actions_dag;
+    if (!filter_dag || query_info.isFinal())
+        return;
+
+    if (!VirtualColumnUtils::isDeterministicInScopeOfQuery(filter_dag->getOutputs().front()))
+        return;
+
+    for (auto iter = stack.rbegin() + 1; iter != stack.rend(); ++iter)
+    {
+        if (auto * filter_step = typeid_cast<FilterStep *>(iter->node->step.get()))
+        {
+            const String & where_condition = filter_dag->getOutputs().front()->result_name;
```

Comment:
as where condition id we prob need smth stronger. at least hash of all filter action dag.